src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs

```diff
@@ -297,6 +219,100 @@ public static string GetFormat(string url)
+                    preparedStream = new MemoryStream();
+                    bufferStream.Position = 0;
+                    await bufferStream.CopyToAsync(preparedStream, 81920, token); // Copy buffered portion
+                    await input.CopyToAsync(preparedStream, 81920, token); // Copy remaining data
+                    preparedStream.Position = 0;
+                }
+            }
+            else
+            {
+                format ??= InspectStreamFormat(input);
+
+                if (!format.Equals(OpenApiConstants.Json, StringComparison.OrdinalIgnoreCase))
+                {
+                    // Buffer stream for non-JSON formats (e.g., YAML) since they require synchronous reading
+                    preparedStream = new MemoryStream();
+                    await input.CopyToAsync(preparedStream, 81920, token);
+                    preparedStream.Position = 0;
+                }
+            }
+
+            return(preparedStream, format);
+        }
+
+        private static string InspectStreamFormat(Stream stream)
+        {
+            if (stream == null) throw new ArgumentNullException(nameof(stream));
+
+            long initialPosition = stream.Position;
+            int firstByte = stream.ReadByte();
+
+            // Check if stream is empty or contains only whitespace
+            if (firstByte == -1)
+            {
+                stream.Position = initialPosition;
+                throw new InvalidOperationException("Stream is empty or contains only whitespace.");
+            }
+
+            // Skip whitespace if present and read the next non-whitespace byte
+            if (char.IsWhiteSpace((char)firstByte))
+            {
+                firstByte = stream.ReadByte();
+
+                // If still whitespace or end of stream, throw an error
+                if (firstByte == -1 || char.IsWhiteSpace((char)firstByte))
+                {
+                    stream.Position = initialPosition;
+                    throw new InvalidOperationException("Stream is empty or contains only whitespace.");
+                }
+            }
+
+            stream.Position = initialPosition; // Reset the stream position to the beginning
+
+            char firstChar = (char)firstByte;
+            return firstChar switch
+            {
+                '{' or '[' => OpenApiConstants.Json,  // If the first character is '{' or '[', assume JSON
+                _ => OpenApiConstants.Yaml             // Otherwise assume YAML
+            };
+        }
+
+        private static string InspectTextReaderFormat(TextReader reader)
+        {
+            // Read the first line or a few characters from the input
+            var input = reader.ReadLine().Trim();
+            return InspectInputFormat(input);
+        }
+
+        private static string InspectInputFormat(string input)
+        {
+            return input.StartsWith("{") || input.StartsWith("[") ? OpenApiConstants.Json : OpenApiConstants.Yaml;
```

Comment:
```suggestion
            return input.StartsWith("{", StringComparison.OrdinalIgnoreCase) || input.StartsWith("[", StringComparison.OrdinalIgnoreCase) ? OpenApiConstants.Json : OpenApiConstants.Yaml;
```

(it'd be better to use the char overload, but since this is still targeting netstandard 2.0, it's not available)